Content.Server/Body/Systems/RespiratorSystem.cs

```diff
@@ -41,155 +56,95 @@ public override void Update(float frameTime)
+        public void Exhale(EntityUid uid, SharedBodyComponent? body=null)
         {
-            return respirator.ProducesGases.ToDictionary(pair => pair.Key, pair => GasProducedMultiplier(respirator, pair.Key, usedAverage));
-        }
-
-        private void ProcessGases(EntityUid uid, RespiratorComponent respirator,
-            BloodstreamComponent? bloodstream,
-            SharedBodyComponent? body)
-        {
-            if (!Resolve(uid, ref bloodstream, ref body, false))
+            if (!Resolve(uid, ref body, false))
                 return;
 
-            var lungs = _bodySystem.GetComponentsOnMechanisms<LungComponent>(uid, body).ToArray();
+            var organs = _bodySystem.GetComponentsOnMechanisms<LungComponent>(uid, body).ToArray();
 
-            var needs = NeedsAndDeficit(respirator);
-            var used = 0f;
+            // exhale gas
 
-            foreach (var (lung, mech) in lungs)
-            {
-                _lungSystem.UpdateLung(lung.Owner, lung, mech);
-            }
+            var ev = new ExhaleLocationEvent();
+            RaiseLocalEvent(uid, ev, false);
 
-            foreach (var (gas, amountNeeded) in needs)
+            if (ev.Gas == null)
             {
-                var bloodstreamAmount = bloodstream.Air.GetMoles(gas);
-                var deficit = 0f;
-
-                if (bloodstreamAmount < amountNeeded)
-                {
-                    // Panic inhale
-                    foreach (var (lung, mech) in lungs)
-                    {
-                        _lungSystem.Gasp((lung).Owner, lung, mech);
-                    }
-
-                    bloodstreamAmount = bloodstream.Air.GetMoles(gas);
-
-                    deficit = Math.Max(0, amountNeeded - bloodstreamAmount);
-
-                    if (deficit > 0)
-                    {
-                        bloodstream.Air.SetMoles(gas, 0);
-                    }
-                    else
-                    {
-                        bloodstream.Air.AdjustMoles(gas, -amountNeeded);
-                    }
-                }
-                else
-                {
-                    bloodstream.Air.AdjustMoles(gas, -amountNeeded);
-                }
-
-                respirator.DeficitGases[gas] = deficit;
-
-                used += (amountNeeded - deficit) / amountNeeded;
+                ev.Gas = _atmosSys.GetTileMixture(Transform(uid).Coordinates);
+                if (ev.Gas == null) return;
             }
 
-            var produced = GasProduced(respirator, used / needs.Count);
-
-            foreach (var (gas, amountProduced) in produced)
+            var outGas = new GasMixture(ev.Gas.Volume);
+            foreach (var (lung, _) in organs)
             {
-                bloodstream.Air.AdjustMoles(gas, amountProduced);
+                _atmosSys.Merge(outGas, lung.Air);
             }
 
-            ClampDeficit(respirator);
+            _atmosSys.Merge(ev.Gas, outGas);
         }
```

Comment:
Is there a reason why you can't just directly merge the gas into ev.Gas and need to use an intermediate outGas?

Reply:
theres not any explicit reason i cant do that but i figured it might be better to reduce the amount of modifications made to the gas that's actually staying around